pkg/jsonutils/jsonutils.go

```diff
@@ -8,63 +8,105 @@ import (
 	"fmt"
 	"io"
 	"os"
-	"strings"
 
 	"github.com/sassoftware/sas-ggdk/pkg/result"
 )
 
 // PrintJSONOn pretty-prints the given data as JSON on the given writer. If the
 // given data is not valid JSON an error is returned.
-func PrintJSONOn(data any, writer io.Writer) error {
-	bites, err := json.MarshalIndent(data, ``, `    `)
-	if err != nil {
+func PrintJSONOn[T any](data T, writer io.Writer) error {
+	bites := result.New(json.MarshalIndent(data, ``, `    `))
+	return result.MapErrorOnly(func(b []byte) error {
+		_, err := fmt.Fprintf(writer, "%s", b)
 		return err
-	}
-	content := string(bites)
-	_, err = fmt.Fprintf(writer, "%s\n", content)
-	return err
+	}, bites)
 }
 
 // ToJSON returns a pretty-printed JSON string for the given data. If the given
 // data is not valid JSON an error is returned. If you are marshaling a large
 // data structure and are concerned about the performance of iteratively growing
 // the destination strings.Builder then create your own builder, grow it to your
-// expected size, and then call PrintJSONOn.
-func ToJSON(data any) result.Result[string] {
-	writer := new(strings.Builder)
-	err := PrintJSONOn(data, writer)
-	if err != nil {
-		return result.Error[string](err)
-	}
-	content := writer.String()
-	return result.Ok(content)
+// expected size, and then call PrintJSONOn. If you want the bytes then call the
+// following.
+//
+//	result.New(json.MarshalIndent(data, ``, `    `))
+func ToJSON[T any](data T) result.Result[string] {
+	bites := result.New(json.MarshalIndent(data, ``, `    `))
+	return result.MapNoError(func(b []byte) string { return string(b) }, bites)
 }
 
-// UnmarshalFromReader inflates the JSON in the given reader and populates the
-// given instance. If the data in the reader is not valid JSON an error is
+// UnmarshalFromReader unmarshals the JSON in the given reader and populates a
```

Comment:
Here and below, it may be worth highlighting that the reader will be read fully before its content is processed.